test/integration/network_extension_discovery_integration_test.cc

```diff
@@ -0,0 +1,707 @@
+  sendDataVerifyResults(5);
+
+  // Wait for configuration expired. Then start a TCP connection.
+  // The missing config network filter will be installed to handle the connection.
+  test_server_->waitForCounterGe(
+      "extension_config_discovery.network_filter." + filter_name_ + ".config_reload", 2);
+
+  IntegrationTcpClientPtr tcp_client = makeTcpConnection(lookupPort(port_name_));
+  auto result = tcp_client->write(data_);
+  if (result) {
+    tcp_client->waitForDisconnect();
+  }
+
+  // Reinstate the configuration.
+  sendXdsResponse(filter_name_, "1", 3);
+  test_server_->waitForCounterGe(
+      "extension_config_discovery.network_filter." + filter_name_ + ".config_reload", 3);
+  sendDataVerifyResults(3);
+}
+
+TEST_P(NetworkExtensionDiscoveryIntegrationTest, BasicSuccessWithTtlWithDefault) {
+  on_server_init_function_ = [&]() { waitXdsStream(); };
+  addFilterChain();
+  addDynamicFilter(filter_name_, false);
+  initialize();
+
+  EXPECT_EQ(test_server_->server().initManager().state(), Init::Manager::State::Initializing);
+
+  // Send 1st config update with TTL 1s, and have network filter drain 5 bytes of data.
+  sendXdsResponse(filter_name_, "1", 5, true);
+  test_server_->waitForCounterGe(
+      "extension_config_discovery.network_filter." + filter_name_ + ".config_reload", 1);
+  sendDataVerifyResults(5);
+
+  // Wait for configuration expired. The default filter will be installed.
+  test_server_->waitForCounterGe(
+      "extension_config_discovery.network_filter." + filter_name_ + ".config_reload", 2);
+  // Start a TCP connection. The default filter drains 2 bytes.
+  sendDataVerifyResults(2);
+}
+
+TEST_P(NetworkExtensionDiscoveryIntegrationTest, BasicFailWithDefault) {
+  on_server_init_function_ = [&]() { waitXdsStream(); };
+  addFilterChain();
+  addDynamicFilter(filter_name_, false, true);
+  initialize();
+
+  EXPECT_EQ(test_server_->server().initManager().state(), Init::Manager::State::Initializing);
+
+  // Send config update with invalid config (bytes_to_drain needs to be >=2).
+  sendXdsResponse(filter_name_, "1", 1);
+  test_server_->waitForCounterGe(
+      "extension_config_discovery.network_filter." + filter_name_ + ".config_fail", 1);
+  // The default filter will be installed. Start a TCP connection. The default filter drain 2 bytes.
+  sendDataVerifyResults(2);
+}
+
+TEST_P(NetworkExtensionDiscoveryIntegrationTest, BasicFailWithoutDefault) {
+  on_server_init_function_ = [&]() { waitXdsStream(); };
+  addFilterChain();
+  addDynamicFilter(filter_name_, false, false);
+  initialize();
+
+  EXPECT_EQ(test_server_->server().initManager().state(), Init::Manager::State::Initializing);
+
+  // Send config update with invalid config (drain_bytes has to >=2).
+  sendXdsResponse(filter_name_, "1", 1);
+  test_server_->waitForCounterGe(
+      "extension_config_discovery.network_filter." + filter_name_ + ".config_fail", 1);
+
+  // New connections will close since there's no valid configuration.
```

Comment:
Is there a way to monitor connections that try to use an invalid configuration?

Reply:
I am not sure there currently is.
I believe we would want a metric similar to listener's ``extension_config_missing``:
https://github.com/envoyproxy/envoy/blob/d2c5c0c9341e16f939f8e85da15850bb2970e244/source/extensions/listener_managers/listener_manager/listener_impl.cc#L958

This stat is increased when a connection is closed due to missing configuration. I intend to add an equivalent to network filters after this one is merged.

Reply:
Ack. That's desired, as otherwise it will be hard to monitor such cases.